Implement auter notification emails

rhodesn opened the pull request:
PR for rackerlabs/auter/issues/75.

Create an auter module for sending notification emails at various stages in the auter patching process. The basic functionality includes adding the package listing, prep/apply errors and reboot timing information to the body of the email.

The module also supports custom user scripts, the output of which will be included in the body of the email.

There are only a few config options at the moment:
- MAIL_TO(default: ""): space separated list of recipients
- MAIL_FROM(default: ""): override the default mail from headers
- MAIL_CC(defaut: ""): space separated list of addresses to CC should the user want mailing lists notified etc
- TRUNCATE_LENGTH(default: 5000): hard limit on character count in the information section of the email, once this limit is reached a warning is logged and no further lines from package lists/errors/custom scripts are added to the emails.

I've also included example pre/post auter scripts, along with a couple example custom scripts.

rhodesn committed:
Initial commit for auter-notify.module

Auter email notification module. Sends a notification email to all
listed recipients in auter-notify.conf. The module comes with a basic
standard email body, however it also accepts custom scripts which can
output content to be added to the mail body.

This also includes example pre/post files for calling the module, and
example custom scripts for creating useful reports on the various
stages.

Remove default TRUNCATE_* values from auter-notify.conf

Left in from testing - these should be empty by default which turns off
truncation.

Reimplement the TRUNCATE functionality in auter-notify.module

Rather than cutting the lines based on individual line character counts,
switch to keeping track of the overall character count for the info
section of the body. Once the watermark has been breached then we trim
the remaining lines and issue a warning.

This requires the output to be broken over several lines, otherwise the
character count will be exceeded on the first line, however it will
still be inserted into the email.

Switch to using uppercase for the variables names

Also fix some shellcheck warnings for variable quoting.

rhodesn committed:
Initial commit for auter-notify.module

Auter email notification module. Sends a notification email to all
listed recipients in auter-notify.conf. The module comes with a basic
standard email body, however it also accepts custom scripts which can
output content to be added to the mail body.

This also includes example pre/post files for calling the module, and
example custom scripts for creating useful reports on the various
stages.

Remove default TRUNCATE_* values from auter-notify.conf

Left in from testing - these should be empty by default which turns off
truncation.

Reimplement the TRUNCATE functionality in auter-notify.module

Rather than cutting the lines based on individual line character counts,
switch to keeping track of the overall character count for the info
section of the body. Once the watermark has been breached then we trim
the remaining lines and issue a warning.

This requires the output to be broken over several lines, otherwise the
character count will be exceeded on the first line, however it will
still be inserted into the email.

Switch to using uppercase for the variables names

Also fix some shellcheck warnings for variable quoting.

reaperzn commented on contrib/email-notifications/auter-notify.conf:
```diff
@@ -0,0 +1,12 @@
+# Auter email notification config file
+# This file should be moved to ${SCRIPTDIR}
```
This file should not be moved to ${SCRIPTDIR} otherwise auter will attempt to execute it. As an example, see https://github.com/rackerlabs/auter/blob/a240f29d1852526cd9b4e00141b362ed916296be/auter.yumdnfModule#L40

rhodesn committed:
Merge remote-tracking branch 'origin/feature/email-notifications' into feature/email-notifications

reaperzn commented on contrib/email-notifications/report-rpmnew.post-apply:
```diff
@@ -0,0 +1,3 @@
+#/bin/bash
+
+[[ `find / -type f -name '*.rpmnew' | wc -l` -gt 0 ]] && find / -type f -name '*.rpmnew' || echo "No new config files"
```
This has potential to cause auter to "hang" for long periods of time. You should rather look at the /var/lib/auter/last-apply-output-default  which should have a list of all the rpmbew and rpmsave files. eg:

```
[root@rhel6-test default]# cat /var/lib/auter/last-apply-output-default
Loaded plugins: fastestmirror
Setting up Install Process
Examining /var/cache/auter/default/auter-0.11-5.el6.noarch.rpm: auter-0.11-5.el6.noarch
Marking /var/cache/auter/default/auter-0.11-5.el6.noarch.rpm as an update to auter-0.9-1.el6.noarch
Loading mirror speeds from cached hostfile
Resolving Dependencies
--> Running transaction check
---> Package auter.noarch 0:0.9-1.el6 will be updated
---> Package auter.noarch 0:0.11-5.el6 will be an update
--> Finished Dependency Resolution

Dependencies Resolved

================================================================================
 Package     Arch         Version          Repository                      Size
================================================================================
Updating:
 auter       noarch       0.11-5.el6       /auter-0.11-5.el6.noarch        42 k

Transaction Summary
================================================================================
Upgrade       1 Package(s)

Total size: 42 k
Downloading Packages:
Running rpm_check_debug
Running Transaction Test
Transaction Test Succeeded
Running Transaction
  Updating   : auter-0.11-5.el6.noarch                                      1/2 
warning: /etc/auter/auter.conf created as /etc/auter/auter.conf.rpmnew
  Cleanup    : auter-0.9-1.el6.noarch                                       2/2 
  Verifying  : auter-0.11-5.el6.noarch                                      1/2 
  Verifying  : auter-0.9-1.el6.noarch                                       2/2 

Updated:
  auter.noarch 0:0.11-5.el6                                                     

Complete!
```

You might replace this line with something like:
```
egrep -o ".*rpmnew" /var/lib/auter/last-apply-output-default | awk '{print $NF}' || echo "No new config files"
```

reaperzn commented on contrib/email-notifications/report-configsnap.post-apply:
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+if [[ -f /root/auter-configsnap-$(date +%Y-%m-%d)/configsnap/post-update.compare ]]; then
```
This entirely depends on configsnap so that should be included in the dependency list. You should also expand your report-configsnap.post-apply section in README.md to include a note that this depends on the specific naming convention for the configsnap `-d` option. You have used the default config from the current configsnap contrib script but this is a customizable option so should be stated in the README.

reaperzn commented on contrib/email-notifications/auter-notify.module:
```diff
@@ -0,0 +1,218 @@
+  function logit() {
+    # If running on a tty, or the --stdout option is provided, print to screen
+    ( tty -s || [[ -n "$STDOUT" ]] ) && echo "$1"
+    logger -p INFO -t auter "$1"
```
This should probably be adjusted to:
```logger -p INFO -t "$(basename "$0")" "$1"```
to avoid confusion in the logs.

reaperzn commented on contrib/email-notifications/auter-notify.module:
```diff
@@ -0,0 +1,218 @@
+function main() {
+  function logit() {
+    # If running on a tty, or the --stdout option is provided, print to screen
+    ( tty -s || [[ -n "$STDOUT" ]] ) && echo "$1"
```
STDOUT is not exported in the main auter script so can't be used here (at least in the context of auter

reaperzn commented on contrib/email-notifications/auter-notify.module:
```diff
@@ -0,0 +1,218 @@
+  fi
+
+  local MAIL_CC
+  MAIL_CC="-c $(sed 's/ / -c /g' <<< "$MAIL_CC")"
```
If no CC address is provided then the variable will simply contain "-c " which will cause the mailx command on line 104 to fail

reaperzn commented on contrib/email-notifications/auter-notify.module:
```diff
@@ -0,0 +1,218 @@
+  if [[ -f "${SCRIPTDIR}/auter-notify.conf" ]]; then
+    source "${SCRIPTDIR}/auter-notify.conf"
+  else
+    logit "ERROR: auter-notify.conf is missing, skipping $0"
```
Should this not exit the script?

reaperzn commented on contrib/email-notifications/auter-notify.module:
```diff
@@ -0,0 +1,218 @@
+  function get_datetime() {
+    date -d "${1:-now}" "+%T %Z"
+  }
+  function run_custom_scripts {
```
No issue here, this is a nice idea. Just make sure you document this in the README.md, especially the fact that this needs to be executable. 

You may also want to add an else clause to at lease log that some files have been skipped.

rhodesn committed:
Updates after PR review

Add configsnap as a dependency for the email-notification contrib
scripts.

Update logit to use `"$(basename "$0")"` rather than `auter` to log the
correct source script.

Use a simple grep in report-rpmnew.post-apply instead of searching the
filesytem for rpmnew files. This should massively speed up the process.

rhodesn committed:
Include dependency on scripts being executable in the README

Addressing the final point raised in the PR - add a note to the README
that all scripts need to be executable otherwise they will be skipped by
either auter or auter-notify.

rhodesn committed:
Merge remote-tracking branch 'upstream/develop' into feature/email-notifications

rhodesn committed:
Updates to formatting/comments/script naming

Rename the pre/post scripts so that they are correctly ordered by bash
when looping through.

Add additional comments to the reporting scripts detailing any
assumptions that are being made for the script to succeed.
